File: problems-extend-adapt/3501-3600/3525_product-remainder-tallies-ii/solution.ts
function remainderTallies(nums: number[], k: number, queries: number[][]): number[] {
    // After the update and the forced prefix removal, the operation picks
    // nums[start..j], so a query counts j >= start whose product from start
    // is x mod k. Each segment tree node stores the counts of its segment's
    // prefix products plus the segment product; merging prepends the left
    // product to the right child's counts, and the suffix query merges the
    // decomposition of nums[start..] left to right while carrying the
    // running product. Every stored value is below k <= 5 and every count
    // below n = 10^5, so int32 arrays hold everything exactly.
    const n = nums.length;
    let size = 1;
    while (size < n) size <<= 1;
    const cnt = new Int32Array(2 * size * k);
    const prod = new Int32Array(2 * size).fill(1);
    for (let i = 0; i < n; ++i) {
        prod[size + i] = nums[i] % k;
        cnt[(size + i) * k + (nums[i] % k)] = 1;
    }
    const merge = (u: number): void => {
        const base = u * k;
        const lrow = 2 * u * k;
        const rrow = lrow + k;
        cnt.set(cnt.subarray(lrow, lrow + k), base);
        const lp = prod[u + u];
        for (let p = 0; p < k; ++p) {
            const c = cnt[rrow + p];
            if (c !== 0) cnt[base + ((lp * p) % k)] += c;
        }
        prod[u] = (lp * prod[u + u + 1]) % k;
    };
    for (let u = size - 1; u >= 1; --u) merge(u);
    const result: number[] = [];
    for (const [index, value, start, x] of queries) {
        const leaf = size + index;
        const row = leaf * k;
        for (let r = 0; r < k; ++r) cnt[row + r] = 0;
        cnt[row + (value % k)] = 1;
        prod[leaf] = value % k;
        for (let u = leaf >> 1; u >= 1; u >>= 1) merge(u);
        let lo = size + start;
        let hi = 2 * size;
        const cur = new Int32Array(k);
        let running = 1;
        while (lo < hi) {
            if (lo & 1) {
                const base = lo * k;
                for (let p = 0; p < k; ++p) {
                    const c = cnt[base + p];
                    if (c !== 0) cur[(running * p) % k] += c;
                }
                running = (running * prod[lo]) % k;
                lo++;
            }
            lo >>= 1;
            hi >>= 1;
        }
        result.push(cur[x]);
    }
    return result;
}

File: problems-extend-adapt/3501-3600/3525_product-remainder-tallies-ii/solution.java
class Solution {

    public int[] remainderTallies(int[] nums, int k, int[][] queries) {
        // After the update and the forced prefix removal, the operation
        // picks nums[start..j], so a query counts j >= start whose product
        // from start is x mod k. Each segment tree node stores the counts
        // of its segment's prefix products plus the segment product;
        // merging prepends the left product to the right child's counts,
        // and the suffix query merges the decomposition of nums[start..]
        // left to right while carrying the running product. Every stored
        // value is below k <= 5 and every count below n, so ints suffice.
        int n = nums.length;
        int size = 1;
        while (size < n) {
            size <<= 1;
        }
        int[] cnt = new int[2 * size * k];
        int[] prod = new int[2 * size];
        java.util.Arrays.fill(prod, 1);
        for (int i = 0; i < n; ++i) {
            prod[size + i] = nums[i] % k;
            cnt[(size + i) * k + (nums[i] % k)] = 1;
        }
        for (int u = size - 1; u >= 1; --u) {
            merge(cnt, prod, u, k);
        }
        int[] result = new int[queries.length];
        for (int qi = 0; qi < queries.length; ++qi) {
            int index = queries[qi][0];
            int value = queries[qi][1];
            int start = queries[qi][2];
            int x = queries[qi][3];
            int leaf = size + index;
            int row = leaf * k;
            for (int r = 0; r < k; ++r) {
                cnt[row + r] = 0;
            }
            cnt[row + (value % k)] = 1;
            prod[leaf] = value % k;
            for (int u = leaf >> 1; u >= 1; u >>= 1) {
                merge(cnt, prod, u, k);
            }
            int lo = size + start;
            int hi = 2 * size;
            int[] cur = new int[k];
            int running = 1;
            while (lo < hi) {
                if ((lo & 1) == 1) {
                    int base = lo * k;
                    for (int p = 0; p < k; ++p) {
                        int c = cnt[base + p];
                        if (c != 0) {
                            cur[(running * p) % k] += c;
                        }
                    }
                    running = (running * prod[lo]) % k;
                    lo++;
                }
                lo >>= 1;
                hi >>= 1;
            }
            result[qi] = cur[x];
        }
        return result;
    }

    private static void merge(int[] cnt, int[] prod, int u, int k) {
        int base = u * k;
        int lrow = 2 * u * k;
        int rrow = lrow + k;
        System.arraycopy(cnt, lrow, cnt, base, k);
        int lp = prod[u + u];
        for (int p = 0; p < k; ++p) {
            int c = cnt[rrow + p];
            if (c != 0) {
                cnt[base + ((lp * p) % k)] += c;
            }
        }
        prod[u] = (lp * prod[u + u + 1]) % k;
    }
}
